src/composer.ts

```diff
@@ -0,0 +1,497 @@
+  }
+
+  /**
+   * Send the transaction group to the network and wait until it's committed to a block. An error
+   * will be thrown if submission or execution fails.
+   *
+   * The composer's status must be SUBMITTED or lower before calling this method, since execution is
+   * only allowed once. If submission is successful, this composer's status will update to SUBMITTED.
+   * If the execution is also successful, this composer's status will update to COMMITTED.
+   *
+   * Note: a group can only be submitted again if it fails.
+   *
+   * @param client - An Algodv2 client
+   * @param waitRounds - The maximum number of rounds to wait for transaction confirmation
+   *
+   * @returns A promise that, upon success, resolves to an object containing the confirmed round for
+   *   this transaction, the txIDs of the submitted transactions, and an array of results containing
+   *   one element for each method call transaction in this group.
+   */
+  async execute(
+    client: Algodv2,
+    waitRounds: number
+  ): Promise<{
+    confirmedRound: number;
+    txIDs: string[];
+    methodResults: ABIResult[];
+  }> {
+    if (this.status === AtomicTransactionComposerStatus.COMMITTED) {
+      throw new Error(
+        'Transaction group has already been executed successfully'
+      );
+    }
+
+    const txIDs = await this.submit(client);
+
+    const firstMethodCallIndex = this.transactions.findIndex((_, index) =>
+      this.methodCalls.has(index)
+    );
+    const indexToWaitFor =
+      firstMethodCallIndex === -1 ? 0 : firstMethodCallIndex;
+    const confirmedTxnInfo = await waitForConfirmation(
+      client,
+      txIDs[indexToWaitFor],
+      waitRounds
+    );
+
+    this.status = AtomicTransactionComposerStatus.COMMITTED;
+
+    const confirmedRound: number = confirmedTxnInfo['confirmed-round'];
+
+    const methodResults: ABIResult[] = [];
+
+    for (const [txnIndex, method] of this.methodCalls) {
+      const txID = txIDs[txnIndex];
+
+      const methodResult: ABIResult = {
+        txID,
+        rawReturnValue: new Uint8Array(),
+      };
+
+      try {
+        if (method.returns.type !== 'void') {
+          const pendingInfo =
+            txnIndex === firstMethodCallIndex
+              ? confirmedTxnInfo
+              : // eslint-disable-next-line no-await-in-loop
+                await client.pendingTransactionInformation(txID).do();
+
+          const logs: string[] = pendingInfo.logs || [];
+
+          // first 4 bytes of SHA-512/256 hash of "return"
+          const returnPrefix = Buffer.from([21, 31, 124, 117]);
+
+          let returnValueEncoded: Buffer | undefined;
+
+          for (let i = logs.length - 1; i >= 0; i--) {
+            const log = Buffer.from(logs[i], 'base64');
+            if (log.byteLength >= 4 && log.slice(0, 4).equals(returnPrefix)) {
+              returnValueEncoded = log.slice(4);
+              break;
+            }
+          }
+
+          if (returnValueEncoded == null) {
```

Comment:
Interesting, is it better to do a loose check here? I think it's because you also want to check for empty slices in addition to the `undefined`, but I'm thinking what would happen for an ABI method with a valid four-byte-only return (maybe an empty tuple?).

Reply:
This if statement will only execute if nothing in the logs array starts with the return prefix. If a method does return an "empty" value by logging only the 4 byte return hash, then `returnValueEncoded` would be a 0 length `Buffer` here, which is not `== null`.